### PR TITLE
Mandd/pareto front enhancement

### LINKED ISSUE #1447: [TASK] Non Dominated front engine
--------
Issue Description
--------
It is needed to develop a method to determine the non-dominated set of points in a generic multi-dimensional space

**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [ ] 1. Is it tagged with a type: defect or task?
- [ ] 2. Is it tagged with a priority: critical, normal or minor?
- [ ] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [ ] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [ ] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [ ] 1. If the issue is a defect, is the defect fixed?
- [ ] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [ ] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [ ] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [ ] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by mandd**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Close #1447

##### What are the significant changes in functionality due to this change request?

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [ ] 1. Review all computer code.
- [ ] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [ ] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [ ] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [ ] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [ ] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [ ] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [ ] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [ ] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

File: framework/utils/frontUtils.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    nondominatedPointMask = np.any(data<data[nextPointIndex], axis=1)
+    nondominatedPointMask[nextPointIndex] = True
+    isEfficient = isEfficient[nondominatedPointMask] 
+    data = data[nondominatedPointMask]
+    nextPointIndex = np.sum(nondominatedPointMask[:nextPointIndex])+1
+  if returnMask:
+    # In this case a True/False mask (isEfficientMask) is returned
+    # Non-dominated points pFront can be obtained as follows:
+    #    mask = nonDominatedFrontier(data,True)
+    #    pFront = data[np.array(mask)]
```
Comment (wangcj05) [inline framework/utils/frontUtils.py:48]: move this commented lines to the docstring of this method

File: framework/utils/frontUtils.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  if returnMask:
+    # In this case a True/False mask (isEfficientMask) is returned
+    # Non-dominated points pFront can be obtained as follows:
+    #    mask = nonDominatedFrontier(data,True)
+    #    pFront = data[np.array(mask)] 
+    isEfficientMask = np.zeros(nPoints, dtype = bool)
+    isEfficientMask[isEfficient] = True
+    return isEfficientMask
+  else:
+    return isEfficient
```
Comment (wangcj05) [inline framework/utils/frontUtils.py:53]: Some explanation how to use isEfficient to get the parrot frontier points, similar to isEfficientMask, and documented in the docstring of this method.

File: framework/utils/frontUtils.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+# External Imports
+import numpy as np
+# Internal Imports
+
+
+
+def nonDominatedFrontier(data, returnMask):
+  """
+    This method is designed to identify the set of non-dominated points (nEfficientPoints)
+    @ In, data, np.array, data matrix (nPoints, nCosts) containing the data point
```
Comment (wangcj05) [inline framework/utils/frontUtils.py:28]: Is the data limited to two dimensions?

**Review by wangcj05 — COMMENTED** [review]
I have some minor comments.

File: framework/utils/frontUtils.py (hunk 1 above)
Comment (mandd) [reply framework/utils/frontUtils.py:48]: moved

File: framework/utils/frontUtils.py (hunk 2 above)
Comment (mandd) [reply framework/utils/frontUtils.py:53]: added as well

File: framework/utils/frontUtils.py (hunk 3 above)
Comment (mandd) [reply framework/utils/frontUtils.py:28]: Yes, we are dealing here with point sets in an N dimensional space where the number of dimensions is equal to nCosts.
As of now the Pareto frontier can deal only with 2 dimensional data (nCosts=2).
This core engine extend the capabilities to an arbitrary number of dimensions.

**Review by mandd — COMMENTED** [review]
Addressing Congjian's comments

**Review by wangcj05 — APPROVED** [review]
This PR is good to me. It may worth to add a unit test for high-dimensional case (i.e. greater than 2D)

**Comment by alfoa** [comment]
Diego, can you check the description of the PR adding the "Close" before the issue id? If you do not do it, the issue will not be closed when this gets merged.

**Comment by mandd** [comment]
> Diego, can you check the description of the PR adding the "Close" before the issue id? If you do not do it, the issue will not be closed when this gets merged.

I always for get it :-P

**Pull request merged**